### ISSUE TITLE
"Set Cookies" not working with firstparty.isolate

### ISSUE THREAD

**Issue by wjtk4444** [stoically/temporary-containers#251]
I cannot get the "Set Cookies" functionality to work. I've tried doing the same in a fresh firefox profile and it didn't work either.
What I'm trying to achieve is to reproduce a functionality of an userscript I wrote that sets a single cookie on youtube website.
```
// ==UserScript==
// @name yt-dark-theme
// @namespace Violentmonkey Scripts
// @match https://www.youtube.com/*
// @grant none
// ==/UserScript==
// 
function setCookie(name, value, expires, path, domain, secure){
	cookieStr = name + "=" + value + "; ";

	if(expires){
		expires = setExpiration(expires);
		cookieStr += "expires=" + expires + "; ";
	}
	if(path){
		cookieStr += "path=" + path + "; ";
	}
	if(domain){
		cookieStr += "domain=" + domain + "; ";
	}
	if(secure){
		cookieStr += "secure; ";
	}

	document.cookie = cookieStr;
}

setCookie("PREF", "f1=50000000&f6=400", false, "/", ".youtube.com", false);
```
Here are the settings I tried using (one of many attempts).
![image](https://user-images.githubusercontent.com/20820296/56871554-2b003b80-6a0f-11e9-8ef1-2c90a9c2cf2a.png)

I suppose I'm doing something wrong. Could You point me in the right direction? As far as I know, there's no guide on how to use this functionality and mozilla's cookies.set() documentation didn't get me anywhere either.

**Comment by stoically** [comment]
I was also using the Set Cookies feature to set dark mode on youtube. You have everything correct in the given example, just needs `www.youtube.com` as Domain Pattern instead (`*.youtube.com` would work too). 

Personally I've switched to using [`invido.us`](https://github.com/omarroth/invidious), though still the official instance instead of my own one. I use a simple [Redirector](https://addons.mozilla.org/firefox/addon/redirector/)

<details>
<summary>rule</summary>

![image](https://user-images.githubusercontent.com/29637501/56888282-e064da80-6a73-11e9-96c5-fe41f984173c.png)
</details>

but, there's also [Invidition](https://addons.mozilla.org/firefox/addon/invidition/) which even replaces embedded yt videos.

Hope that helps. Let me know if not.

**Comment by wjtk4444** [comment]
Seems like it was the domain pattern's fault, but I have noticed another thing.
"Set Cookies" doesn't work with `privacy.firstparty.isolate` set to `true`. Should I open a new issue? Is this even fixable?

> Personally I've switched to using invido.us

I know about this one, but I only use youtube as a search engine and play videos in mpv instead. As long as the videos themselves are hosted on google servers I see no benefit to using it over youtube.
> I use a simple Redirector

Thanks for this one, seems to react faster than my userscripts when it comes to redirections, and I do quite a few of those. Same thing goes for the "Set Cookies" functionality - youtube isn't the only place where I'd like to use it.

**Comment by stoically** [comment]
> doesn't work with privacy.firstparty.isolate set to true

Ah, yeah, needs `firstPartyDomain` as [additional parameter to `cookies.set`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/cookies/set#Parameters). I'll look into adding that (and `sameSite` while I'm at it).

> As long as the videos themselves are hosted on google servers

It's possible to enable the "proxy" option on invidious instances (and Invidition does that by default by passing `&local=true`), which makes the instance connect to googlevideos instead of the client.

**Comment by wjtk4444** [comment]
> `&local=true`

Sadly, it seems that `youtube-dl`'s [extractor](https://github.com/ytdl-org/youtube-dl/blob/master/youtube_dl/extractor/youtube.py) treats invidio the exact same way as regular youtube and several other frontends.

**Comment by stoically** [comment]
Not sure what you mean by that?

**Comment by wjtk4444** [comment]
[youtube-dl](https://github.com/ytdl-org/youtube-dl/), the program that [mpv](https://github.com/mpv-player/mpv) uses as a backend to fetch videos from websites doesn't have a separate extractor for invidio and youtube. They use the same one, connecting directly to google servers. In order to use the "proxy" option that invidio offers, I would have to either resort to watch videos in a web browser or manually extract a direct url to the playable resource and feed it to the player.

>Not sure what you mean by that?

It's just me complaining about things. Sorry for the off-topic.

**Comment by stoically** [comment]
I see. So, if you feed youtube-dl the url with `&local=true` attached, it *should* connect to the invidious instance instead of googlevideos, because the hls m3u8 playlists *should* contain urls to the instance, not google. If that's not the case, that'd be a candidate for an issue over at invidious.

**Comment by wjtk4444** [comment]
>if you feed youtube-dl the url with &local=true attached, it should connect to the invidious instance instead of googlevideos

That's the expected behavior from the user's point of view, but if you take a look at the [youtube extractor code](https://github.com/ytdl-org/youtube-dl/blob/master/youtube_dl/extractor/youtube.py#L347) (and notice that there's not a separate extractor for invidio), it becomes clear that it only grabs video id and then does it's job by using youtube API, completely ignoring the domain name of the url it was feeded. So, it's not really an issue with either youtube-dl or invidio, just a lazy design of an extractor. Which could be improved. 
https://github.com/ytdl-org/youtube-dl/issues/18859#issuecomment-487655197

**Comment by stoically** [comment]
Ah, understood. Thanks for the clarification!

**Comment by stoically** [comment]
Btw, just realized that I linked to Redirector above, but I'm actually using [Request Control](https://addons.mozilla.org/firefox/addon/requestcontrol/) (and the screenshot is from there too).

**Comment by wjtk4444** [comment]
Cool, it's basically a mini-firewall inside of firefox. It has it's [problems](https://github.com/tumpio/requestcontrol/issues/98) though. 
Are my use cases really so weird that nothing works ootb? Well, at least here goes the power of (F)OSS - I can write issues and hope for them to get addressed some day.

**Comment by stoically** [comment]
Interesting - I wonder why/how Redirector works in this case. On-site navigations using `script` (which essentially dynamically load content via XHR / fetch) shouldn't be redirectable at all.

**Comment by wjtk4444** [comment]
>shouldn't be redirectable at all.

More like 

>shouldn't be ~~redirectable~~ at all.

I have no idea why anyone would cripple their sites like that.

> I wonder why/how Redirector works in this case.

See [my thoughts](https://github.com/tumpio/requestcontrol/issues/98#issuecomment-488082300) on the topic.

**Comment by stoically** [comment]
Well, dynamically loading content over xhr in itself isn't so bad, since it prevents the whole page from reloading. Though, the author of Request Control has a point, since the history state is changed, it'd probably be possible to listen to [onHistoryStateUpdated](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webNavigation/onHistoryStateUpdated) and make that redirectable.

**Comment by wjtk4444** [comment]
Sorry for being a nag, but are there any chances for this enhancement to land in the 
next release? 

> Ah, yeah, needs `firstPartyDomain` as [additional parameter to `cookies.set`](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/cookies/set#Parameters). I'll look into adding that (and `sameSite` while I'm at it).

According to Your comment, it looks like it's (only?) a matter of adding an additional parameter to [cookie.js#L61](https://github.com/stoically/temporary-containers/blob/master/src/background/cookies.js#L61). I'd try it myself, but I don't really know how to build an addon. Also, running a custom-built version would prevent me from updating in the future.

**Comment by stoically** [comment]
No need to build the Add-on, just modify the `src/` and [load the Add-on temporarly](https://github.com/stoically/temporary-containers#run-in-firefox).

Published version 0.99 with the enhancement.

**Comment by wjtk4444** [comment]
> No need to build the Add-on, just modify the `src/` and [load the Add-on temporarly](https://github.com/stoically/temporary-containers#run-in-firefox).

Thanks, good to know that.

There's one (hopefully last) problem - I still can't get the cookies to be set with `fpi` enabled. 

![image](https://user-images.githubusercontent.com/20820296/59853138-cf439600-935f-11e9-9278-05c29e75d0f2.png)

When I disable both `fpi` **AND** clear the `firstPartyDomain` field it works just like it used to before the 0.99 update.
I believe that I tried every single combination possible for `firstPartyDomain` and `sameSite` for youtube (and a few other sites) but I couldn't get it to set a single cookie with `fpi` enabled. 

I used a fresh firefox profile (67.0.2), with only Temporary Containers installed and `privacy.firstparty.isolate` being the only prerefence I played with.
